### PR TITLE
Fix AWS NLB security group updates

### LINKED ISSUE #60825: Failure to update security groups for NLBs after rolling update
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

After a rolling update of the cluster, NLB kept failing health checks because security group did not have rules for new EC2 instance.

**What you expected to happen**:

Security group should have been updated with both TCP rules for NLBs to work correctly.

**How to reproduce it (as minimally and precisely as possible)**:

- Create a service object with type LoadBalancer and adding the annotation for NLBs. FWIW, I had 6 services and each had their own NLBs. It’s possible that the bug happens when there are multiple NLBs causing multiple AWS updates.
- Perform a rolling update of the instance group. This causes the instances to be terminated and new ones to be brought up. AFAICT, this should result in two changes to the security group per EC2 instance. When the instance is terminated, the rules for that nodePort should be deleted. When the new node is created, the rule should be added again.
- However, after the rolling update, the rule doesn’t get added. As a result, the NLB’s health checks fail.

**Anything else we need to know?**:

- As mentioned earlier, this happened on a cluster where there were multiple NLBs.
- This doesn’t seem to be a deterministic failure. 
- The `kube-controller` log has the following error message:
```
I0305 18:04:40.968739       6 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"kafka-1", Name:"outside-4", UID:"805bbe82-2033-11e8-ab46-06d9f670eba0", APIVersion:"v1", ResourceVersion:"6076", FieldPath:""}): type: 'Warning' reason: 'LoadBalancerUpdateFailed' Error updating load balancer with new hosts [ip-172-20-124-189.us-west-2.compute.internal ip-172-20-120-186.us-west-2.compute.internal ip-172-20-51-19.us-west-2.compute.internal ip-172-20-45-228.us-west-2.compute.internal ip-172-20-72-38.us-west-2.compute.internal ip-172-20-105-78.us-west-2.compute.internal ip-172-20-69-148.us-west-2.compute.internal ip-172-20-88-181.us-west-2.compute.internal ip-172-20-38-132.us-west-2.compute.internal ip-172-20-103-250.us-west-2.compute.internal ip-172-20-50-106.us-west-2.compute.internal ip-172-20-81-141.us-west-2.compute.internal ip-172-20-55-251.us-west-2.compute.internal ip-172-20-65-146.us-west-2.compute.internal ip-172-20-103-197.us-west-2.compute.internal]: InvalidPermission.NotFound: The specified rule does not exist in this security group.
```

- I manually edited the Service object to see if it would simply retrigger an update on the cloud resources. It actually did. The rules got added to the security group.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-09T21:51:54Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"clean", BuildDate:"2018-01-04T11:40:06Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
AWS

- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
kops 1.8.1
- Others:

### PR DESCRIPTION
**Pull request by kellycampbell**
This corrects a problem where valid security group ports were removed
unintentionally when updating a service or when node changes occur.

**What this PR does / why we need it**:

There is a bug in the existing logic that causes valid security group port mappings to be removed incorrectly.

**Which issue(s) this PR fixes**:

Fixes #60825

**Special notes for your reviewer**:

This still needs some unit tests. The existing loadbalancer tests are very minimal. Any pointers to how to mock security group actions?

**Release note**:

```release-note
Fix AWS NLB security group updates where valid security group ports were incorrectly removed
when updating a service or when node changes occur.
```

Root cause was an assumption that the list of security groups was actually a set.

Here's some psuedo code of the cause:

```
existingGroups = [g1:80, g1:80]  // <-- NOT a Set
desiredGroups = [g1:80]

portChanges = { desiredGroups : ports }
// portChanges = [g1:80 = true]

for g in existingGroups:
    desiredSet.Insert(portChanges[g] == true)
    // iter1: desiredSet = [ g1:80 ]
    // iter2: desiredSet = [ ]
    existingSet = portsForGroup() 
    // i1: existingSet = [ g1:80 ]
    // i2: existingSet = [ g1:80 ]

    alreadyExists = desiredSet.Intersection(existingSet)
    removeAlreadyExistsFromPortChanges()
    // i1: portChanges = []
    // i2: portChanges = []

    notDesired = existingSet.Minus(desiredSet)
    // i1: notDesired = []
    // i2: notDesired = [g1:80]
    setPortChangeToFalseForNotDesired
```

### CODE REVIEW

**Comment by kellycampbell** [comment]
Really simple test case that reproduced the issue easily for me:

1. Create two services with NLB load balancer - SG's get update to allow the node ports to 0.0.0.0/0
2. Alter one of the services e.g. by adding a new label in the metadata
3. Apply the alteration. The old controller logic removes its node port from the SG.
4. You can continue toggling the SG just by applying label updates to the service.

**Comment by rosskukulinski** [comment]
@kris-nova related to 1.12/1.13 AWS goals: https://github.com/kubernetes/features/issues/423#issuecomment-407512634

**Review by geojaz — APPROVED** [review]
/lgtm

**Comment by kellycampbell** [comment]
@Freyert thanks for the unit tests! You'll need to sign the kubernetes CLA. See instructions above.

**Comment by justinsb** [comment]
This LGTM - I think it needs a gofmt though.  Applying hold so that @micahhausler can have a chance to look, but I think it's a pretty clear fix now that I see it :-)

/assign @micahhausler 

/approve
/lgtm
/hold

**Comment by justinsb** [comment]
Ah - I also see it lost "cncf-cla:yes" because of the commit from @Freyert .  @Freyert are you able to sign the CLA?

**Comment by kellycampbell** [comment]
I fixed the gofmt and rebased.

**Comment by Freyert** [comment]
Was having some issues following the CLA steps, but just got it done. Test failures seem like they are unrelated, but at the same time not quite?

**Review by micahhausler — APPROVED** [review]
LGTM Thanks!

**Comment by Freyert** [comment]
`pull-kubernetes-integration` is flakey -> #69376

**Comment by Freyert** [comment]
@justinsb @micahhausler just checking again what needs to be done to remove the `hold`.

**Comment by Freyert** [comment]
What's the procedure with merging this? I see it needs the `lgtm` label again, but does someone have to push the button afterward?

**Comment by acaire** [comment]
Thanks @kellycampbell this is great. We're running private clusters in VPCs with multiple CIDRs and the health check ingress rules only factor in the main CIDR, not the CIDR for the relevant target group instances. I'm still stepping through the code to see where the CIDRs originate from, just wanted to give you a heads up as I was hoping to figure out a fix before this is merged.

**Comment by kellycampbell** [comment]
/label priority/important-soon

@micahhausler I'd like to get this bugfix into the v1.13 milestone and release if possible.

Also I would like to work on backporting it to 1.12 and 1.11 since NLB doesn't work at all for us without this and we'd like to use one of the features NLB provides.

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: Is there an `else` block missing here?

OK to do in a follow-on PR

Comment (kellycampbell) [reply pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: Yeah, that looks like it was overlooked in the original code from @micahhausler https://github.com/kubernetes/kubernetes/commit/f9445b9dc7a79b3f96ab5e4086b46e731a6108b7#diff-b390ae3de185e9f13a631a5c07b8f3ffR646

**Comment by brenix** [comment]
Did this get cherry-picked to the 1.11 release? I see the cherry-pick for 1.12, but not 1.11. It would be nice if it was as the latest release kops supports is 1.11

**Comment by kellycampbell** [comment]
> Did this get cherry-picked to the 1.11 release? I see the cherry-pick for 1.12, but not 1.11. It would be nice if it was as the latest release kops supports is 1.11

I don't think this has been picked into 1.13 yet either.

**Comment by mbelang** [comment]
@M00nF1sh In which 1.11 version it has been release. It is a bit hard to follow. From what I see it has been released in 1.11.7? is that right?

**Comment by FrederikNS** [comment]
Yes, it was released in 1.11.7, however 1.11.8 contains some important security fixes, so I would recommend upgrading to that

**Pull request merged**
